inspirehep/modules/tools/authortool.py

```diff
@@ -0,0 +1,311 @@
+
+re_emptyline = re.compile(r'\n\s*\n', re.UNICODE)
+re_hyphens = re.compile(\
+    r'(\\255|\u02D7|\u0335|\u0336|\u2212|\u002D|\uFE63|\uFF0D)', re.UNICODE)
+re_multiple_space = re.compile(r'\s{2,}', re.UNICODE)
+re_potential_key = re.compile(r"^(?:\d|[^\w.'-])+$", re.UNICODE)
+re_trailing_nonword = re.compile(r"(?:\d|[^\w.'-])+$", re.UNICODE)
+re_symbols = re.compile(r'\W+', re.UNICODE)
+re_initial_wout_dot = re.compile(r'^[A-Z]$', re.UNICODE)
+
+line_pattern_single = {'alpha':re.compile(r'^([a-z]+)\.*$', re.UNICODE),
+                       'digit':re.compile(r'^(\d+)\.*$', re.UNICODE),
+                       'symbol':re.compile(r'^(.)\.*$', re.UNICODE)}
+
+line_pattern_line = {'alpha': re.compile(r'^([a-z]+)[ .]+(.*)', re.UNICODE),
+                     'digit': re.compile(r'^(\d+)[ .]*(.*)', re.UNICODE),
+                     'symbol': re.compile(r'^(.)[ .]+(.*)', re.UNICODE)}
+
+def clean_word(word):
+    """
+    Separate potential aff-ids from potential name.
+    E.g.: Ali*12% -> Ali, ['*', '12' '%']
+    """
+
+    aff_ids = []
+    nonwords = re_trailing_nonword.search(word)
+    if nonwords:
+        nonword = nonwords.group(0)
+        word = re_trailing_nonword.sub('', word)
+        symbols = re_symbols.findall(nonword)
+        if symbols:
+            aff_ids += symbols
+            for rest in re_symbols.split(nonword):
+                if rest:
+                    aff_ids.append(rest)
+        else:
+            aff_ids.append(nonword)
+    return word, aff_ids
+
+def parse_authors(text, affiliations):
+    """
+    Parse author names and convert to Lastname, Firstnames.
+    Can be separated by ',', newline or affiliation tag.
+    Return list of (author_fullname, [author_affiliations])
+    """
+    authors = []
+
+    text = text.replace(' and ', ',')
+    text = text.replace('\n', ',')
+    text = text.replace(',', ' , ')
+    text = re_hyphens.sub('-', text)
+    text = re_multiple_space.sub(' ', text)
+    text = text.strip()
+
+    aff_keys = affiliations.keys()
+    key_type = ''
+    if aff_keys:
+        if aff_keys[0].isalpha():
+            key_type = 'alpha'
+        elif aff_keys[0].isdigit():
+            key_type = 'digit'
+        else:
+            key_type = 'symbol'
+
+    author_names = []
+    author_affs = []
+    fullname = ''
+    for word in text.split(' '):
+        if word in aff_keys or word == ',' or re_potential_key.search(word):
+            # author name stops here, add affiliations
+            if author_names:
+                if len(author_names) > 1:
+                    fullname = '%s, %s' % \
+                        (author_names[-1], ' '.join(author_names[:-1]))
+                else:
+                    fullname = author_names[0]
+                fullname = fullname.replace('. ', '.')
+                author_names = []
+            if word in aff_keys:
+                author_affs.append(affiliations[word])
+            elif not word == ',':
+                warnings.warn(
```

Comment:
you should probably be raising exceptions instead of warnings. The former will be caught and shown to the user, the latter might end up in some webserver logs at best.

Reply:
Then the user has nothing.
It's better to have something and fix the mistakes manually.

I was hoping it's possible to display the warnings the same way errors are now returned + the result.

Reply:
Otherwise it might be better to ignore the warnings.

Reply:
I asked @jacquerie and there doesn't seem to be a way to display warnings. Exceptions are shown to the user, who can correct them and try again. I would say that's better than silently going on and maybe lose some authors.

Reply:
too bad ...
If I parse something like

    P.Carniti a, b L. di Caprio b a  M. Faverzani b
    a  A.Giachero bb, a 
    
    a
    INFN
    b
    Universita di Roma
    bb
    Universita di Milano-Bicocca
    
It works OK, but I would like to alert the cataloger to have a closer look whether 'di' is part of a name or not. Throwing an error without the result is not an option. What if one author really has no affiliation?

You know more  about the integration.
What if I write a new function that returns both the result-record and some text?
The record can be used to pre-fill the editor. And in the current web-interface you would first print the text followed by the record. Is that feasible?

Reply:
> What if I write a new function that returns both the result-record and some text?

Yes, you have to return something like that. For example you can return an object like
```JSON
{
    "authors": [
        {
            "affiliations": [
                "INFN",
                "Universita di Roma"
            ],
            "full_name": "Carniti, P."
        },
        ...
    ],
    "warnings": [
        {"message": "Author number 2 contains critical word 'di'. Please check if it's actually part of a name."},
        ...
    ]
}
```